lib/FrontendTool/ReferenceDependencies.cpp

```diff
@@ -22,90 +22,156 @@
 #include "swift/AST/NameLookup.h"
 #include "swift/AST/ReferencedNameTracker.h"
 #include "swift/AST/Types.h"
-#include "swift/Frontend/FrontendOptions.h"
 #include "swift/Basic/LLVM.h"
+#include "swift/Frontend/FrontendOptions.h"
+#include "swift/Frontend/ReferenceDependencyKeys.h"
 #include "llvm/ADT/MapVector.h"
 #include "llvm/ADT/SetVector.h"
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/YAMLParser.h"
 
 using namespace swift;
+using namespace reference_dependency_keys;
+
+namespace {
+/// Emits the reference dependencies from the frontend so that the driver
+/// can compute a dependency graph for the whole module, and use it to decide
+/// which files need to be recompiled when doing incremental compilation.
+class ReferenceDependenciesEmitter {
+  SourceFile *const SF;
+  const DependencyTracker &depTracker;
+  llvm::raw_ostream &out;
+
+  ReferenceDependenciesEmitter(SourceFile *const SF,
+                               const DependencyTracker &depTracker,
+                               llvm::raw_ostream &out)
+      : SF(SF), depTracker(depTracker), out(out) {}
+
+public:
+  /// Emits the provided and depended-upon dependencies to a file
+  ///
+  /// \param diags Where problems opening the file are emitted
+  /// \param SF The SourceFile containing the code with the dependences
+  /// \param depTracker The entities depended-upon
+  /// \param outputPath The path of the file where the dependencies are written
+  /// to \return true on error
+  static bool emit(DiagnosticEngine &diags, SourceFile *SF,
+                   const DependencyTracker &depTracker, StringRef outputPath);
+
+  /// Emit the dependencies.
+  static void emit(SourceFile *SF, const DependencyTracker &depTracker,
+                   llvm::raw_ostream &out);
+
+private:
+  /// Opens file for reference dependencies. Emits diagnostic if needed.
+  ///
+  /// \return nullptr on error
+  static std::unique_ptr<llvm::raw_fd_ostream> openFile(DiagnosticEngine &diags,
+                                                        StringRef OutputPath);
+  /// Emits all the dependency information.
+  void emit() const;
+
+  void emitProvides() const;
+  void emitDepends() const;
+  void emitInterfaceHash() const;
+};
+
+/// Emits the declarations provided by a source file.
+class ProvidesEmitter {
+  const SourceFile *const SF;
+  llvm::raw_ostream &out;
+
+  ProvidesEmitter(const SourceFile *const SF, llvm::raw_ostream &out)
+      : SF(SF), out(out) {}
+
+public:
+  /// Emit declarations
+  ///
+  /// \param SF Contains the declarations to emit
+  ///
+  /// \param out Where the declarations are emitted
```

Comment:
This doc comment isn't adding any information (the whole thing, not just this line).